src/client/CodeEditor/getOrCreateEditor.ts

```diff
@@ -195,6 +198,69 @@ export const getOrCreateEditor = ({
 
   extensions.push(rotationIndicator);
 
+/*
+  extensions.push(autocompletion({
+    override: [tsComplete]
+  }))*/
+
+  // displays autocompletes
+  async function tsComplete(ctx) {
+
+    const compilerOptions = {};
+    const fsMap = await tsvfs.createDefaultMapFromCDN(
+      compilerOptions,
+      ts.version,
+      true,
+      ts
+    );
+    fsMap.set(name, text);
+    const system = tsvfs.createSystem(fsMap);
+    const env = tsvfs.createVirtualTypeScriptEnvironment(
+      system,
+      [name],
+      ts,
+      {
+      }
+    );
+
+    let tsCompletions = env.languageService.getCompletionsAtPosition(
+      name,
+      ctx.pos,
+      {}
+    );
+
+    if (!tsCompletions) return { from: ctx.pos, options: [] };
+
+    const Text = ctx.state.doc.toString();
+
+    let lastWord, from;
+    for (let i = ctx.pos - 1; i >= 0; i--) {
+      if ([' ', '.', '\n', ':', '{'].includes(Text[i]) || i === 0) {
```

Comment:
It's worth adding a comment here describing what this cryptic code is doing. If it's copy-pasted from somewhere, best to include the URL to the upstream code in a comment.